### PR TITLE
add test for reading phases from nlloc v7 hyp file

### LINKED ISSUE #3223: writing catalog to NLL obs doesn't handle 12 (etc) channels properly
### Avoid duplicates

- [X] I searched existing issues

### Bug Summary

Noticing that when one exports a catalog object to NLL's obs format, it converts any channel outside of `ZNEH` to a `?`. Not really an issue as NLL still handles it fine.

e.g. obspy/io/nlloc line 400 or so
```
    for pick in catalog[0].picks:
        wid = pick.waveform_id
        station = wid.station_code or "?"
        component = wid.channel_code and wid.channel_code[-1].upper() or "?"
        if component not in "ZNEH":
            component = "?"
        onset = ONSETS_REVERSE.get(pick.onset, "?")
        phase_type = pick.phase_hint or "?"
        polarity = POLARITIES_REVERSE.get(pick.polarity, "?")
```

However when reading NLL's output `.hpy` file back in, this (presumably horizontal) channel is now converted to `N` which then becomes a problem if the station's actual channels are `12`.

It seems like a trivial fix to change ` if component not in "ZNEH"` to at least  `ZNE123H` (or even `ZNE123HABCUVWTR`?) unless I am missing something. I know NLL will handle 123 but not completely sure about the more esoteric channels. 

### Code to Reproduce

_No response_

### Error Traceback

_No response_

### ObsPy Version?

1.3.1.post0+390.g3ad5eee621.obspy.master

### Operating System?

_No response_

### Python Version?

_No response_

### Installation Method?

developer installation / from source / git checkout

### PR DESCRIPTION
**Pull request by megies**
### What does this PR do?

Fixes issues outlined in #3223 
 - phase lines are different in new nlloc versions, items regarding Arrivals are shifted one index to the right
 - avoid replacing component code with "?", actually just write full channel code unchanged in nlloc "COMP[ONENT]" field

### Why was it initiated?  Any relevant Issues?

fixes #3223 

### PR Checklist
- [ ] Correct base branch selected? `master` for new features, `maintenance_...` for bug fixes
- [ ] This PR is not directly related to an existing issue (which has no PR yet).
- [ ] While the PR is still work-in-progress, the `no_ci` label can be added to skip CI builds
- [ ] If the PR is making changes to documentation, docs pages can be built automatically.
      Just add the `build_docs` tag to this PR.
      Docs will be served at [docs.obspy.org/pr/{branch_name}](https://docs.obspy.org/pr/) (do not use master branch).
      Please post a link to the relevant piece of documentation.
- [ ] If all tests including network modules (e.g. `clients.fdsn`) should be tested for the PR,
      just add the `test_network` tag to this PR.
- [ ] All tests still pass.
- [ ] Any new features or fixed regressions are covered via new tests.
- [ ] Any new or changed features are fully documented.
- [ ] Significant changes have been added to `CHANGELOG.txt` .
- [ ] First time contributors have added your name to `CONTRIBUTORS.txt` .
- [ ] If the changes affect any plotting functions you have checked that the plots
      from all the CI builds look correct. Add the "upload_plots" tag so that plotting 
      outputs are attached as artifacts. 
- [ ] New modules, add the module to `CODEOWNERS` with your github handle
- [ ] Add the `ready for review` tag when you are ready for the PR to be reviewed.

### CODE REVIEW

**Comment by megies** [comment]
@filefolder this should do it, let me know if there's anything else you encountered

**Comment by megies** [comment]
CC @trichter FYI, I just noticed you did something with nlloc file formats in git history

**Comment by megies** [comment]
I'll go ahead and merge this now, we can always tweak later if needed. Wanna get a 1.4.0rc1 tagged..

**Comment by trichter** [comment]
Tests are passing. Hoping they cover all circumstances. I remember some "?"s in test files. Or was it another format? 8O

**Comment by megies** [comment]
Yeah, using `?` as filler for empty fields, since its free form

**Comment by filefolder** [comment]
sorry I couldn't test immediately (i think our time zones are perfectly out of phase) but I think there is still a glitch

```
v2
PHASE ID    Ins Cmp   On Pha  FM   Date    HrMn   Sec     Err  ErrMag    Coda      Amp       Per       PriorWt  >   TTpred    Res       Weight    StaLoc(X  Y           Z)      SDist    SAzim  RAz  RDip RQual   Tcorr     TTerr
NWAO         ?    Z    ? P      ? 20221031 0502   41.3602 GAU -1.00e+00 -1.00e+00 -1.00e+00 -1.00e+00    1.0000 >   12.7040   -0.3041    1.2104  -24.3880 -103.3008   -0.3800   73.3529 107.42  40.0 200.0  0     0.0000    0.6352
NWAO         ?    ?    ? S      ? 20221031 0502   51.4602 GAU -1.00e+00 -1.00e+00 -1.00e+00 -1.00e+00    1.0000 >   22.7013   -0.2013    1.1613  -24.3880 -103.3008   -0.3800   73.3529 107.42  40.0 200.0  0     0.0000    1.1351
KLBR         ?    Z    ? P      ? 20221031 0502   56.8602 GAU -1.00e+00 -1.00e+00 -1.00e+00 -1.00e+00    1.0000 >   27.3100    0.5900    1.0455   24.1420   45.4457   -0.3200  173.5594  43.07  40.0 200.0  0     0.0000    1.3655
KLBR         ?    N    ? S      ? 20221031 0503   19.2602 GAU -1.00e+00 -1.00e+00 -1.00e+00 -1.00e+00    1.0000 >   48.6028    1.6972    0.7997   24.1420   45.4457   -0.3200  173.5594  43.07  40.0 200.0  0     0.0000    2.4301
RKGY         ?    Z    ? P      ? 20221031 0503    2.0002 GAU -1.00e+00 -1.00e+00 -1.00e+00 -1.00e+00    1.0000 >   32.6302    0.4098    1.3358  -47.9065 -290.5957   -0.0300  214.3488 167.48  40.0 200.0  0     0.0000    1.6315
RKGY         ?    N    ? S      ? 20221031 0503   25.8802 GAU -1.00e+00 -1.00e+00 -1.00e+00 -1.00e+00    1.0000 >   57.7240   -0.8041    1.0402  -47.9065 -290.5957   -0.0300  214.3488 167.48  40.0 200.0  0     0.0000    2.8862

```
in V2 / NLL7 currently arrival `azimuth` is pointing to `RAz` but it should be pointing to `SAzim` for `station_ azimuth` (https://docs.obspy.org/packages/autogen/obspy.core.event.origin.Arrival.html#obspy.core.event.origin.Arrival). so the index should be one less and actually the same index as in V1, `line[23]`

also, `takeoff_angle` is pointing to `RDip` (here 200) but this seems wrong or possibly mislabeled in NLL as I am not sure how a dip could have that value. The other option, `RAz` (here 40) seems more reasonable. However given that _every_ arrival for this event has the same 200/40 values for these I am skeptical about their accuracy...I would probably just revert back to the original behaviour of not writing them. I do like that it writes the `Weight` now though!

also also, looking though older issues (https://github.com/obspy/obspy/issues/2371) I am not sure that the default `ErrMag` should be -1 as this appears to _actually_ be read in as 1 second? It seems like this was fixed in 2019 but something may have happened to revert it. IMO a perfectly reasonable default would be 0.1 but even 0 would probably be better. In theory this value should be superseded by `LOCGAU` in one's NLL conf file regardless, assuming it's set.. .

**Comment by megies** [comment]
> sorry I couldn't test immediately (i think our time zones are perfectly out of phase) but I think there is still a glitch

no worries and thanks for looking at it again

> in V2 / NLL7 currently arrival `azimuth` is pointing to `RAz` but it should be pointing to `SAzim` for `station_ azimuth` (https://docs.obspy.org/packages/autogen/obspy.core.event.origin.Arrival.html#obspy.core.event.origin.Arrival). so the index should be one less and actually the same index as in V1, `line[23]`

I think on old format it is parsing `Raz` too.

QuakeML docs for `Arrival.azimuth` say:

```
azimuth       Azimuth of station as seen from the epicenter.
```

I am inclined to follow your suggestion, I kind of agree that this might rather mean a pure geometrical azimuth in terms of great circle connection rather than a ray takeoff angle.

**I will change it to use `Sazim` for both old and new format.**

> also, `takeoff_angle` is pointing to `RDip` (here 200) but this seems wrong or possibly mislabeled in NLL as I am not sure how a dip could have that value. The other option, `RAz` (here 40) seems more reasonable. However given that _every_ arrival for this event has the same 200/40 values for these I am skeptical about their accuracy...I would probably just revert back to the original behaviour of not writing them. I do like that it writes the `Weight` now though!

It looks like your location run might have some issues? If I read that file correctly your hypocenter ended up above 0 (which I assume is mean sea level, judging from your GRID specification, which seems to go from -1.5 to about 50  -- km?)? In any case [in nonlinloc there are some checks done](https://github.com/alomax/NonLinLoc/blob/700ed91a2f94d6996a2baa679475a6767dac137b/src/NLLocLib.c#L1248-L1254) after calculating that ray dip and if that dip is not in between 0 and 180, the quality of the ray takeoff calculations is set to 0 (NLLoc docs: `Quality of take-off angle estimation (0 = unreliable, 10 = best) `). It looks safe to assume that if that quality field is 0 we might want to skip reading those values. I actually am on NLLoc 7 myself and redid a location run with 9 stations and the values for these fields look fine in my case and `RQual` is `9`.

```C
        if (iavailable < 0) { // angles grids not available
            pfmarr->first_mot_quality = 0.0;
        }
        // check some things
        if (pfmarr->ray_azim < 0.0 || pfmarr->ray_azim > 360.0 || pfmarr->ray_dip < 0.0 || pfmarr->ray_dip > 180.0) {
            pfmarr->first_mot_quality = 0.0;
        }
```

**Long story short, I think using this `RDip` field is fine, but will add skipping to read it if quality is `0`**



> edit: also FWIW issue #2371 can be closed as I see that the default output `ErrMag` is 0.1

Not sure what `ErrMag` is or what it has to do with the ticket you linked, but I see that that ticket #2371 still needs to be taken care of so I'll do it here in a second


> edit2: there is also issue #2831 where the amplitude is not written to the OBS file. i'm not really sure what this "does" necessarily (is anyone calculating magnitudes in NLL?) but now seems like as good a time as any

I think this would need some looking into first, I'll leave that one unhandled for now

**Comment by filefolder** [comment]
ahhh I see, yes, my example was a bad (?) one with the hypocenter slightly above sea level. in deeper events the values are more realistic. disabling output if quality is 0 is a great workaround. 

`ErrMag` is just the default pick uncertainty corresponding to `LOCGAU` I believe. it looked like in 2019 obspy was writing `NLLOC_OBS` with this value set to  -1, which effectively became 1. I can confirm now that obspy writes this out to 0.1 as default, which is probably the sanest option, so case closed on that ticket.

**Comment by megies** [comment]
> I can confirm now that obspy writes this out to 0.1 as default, which is probably the sanest option, so case closed on that ticket.

huh.. it looked like to me we were still writing out -1, let me look at it again. I just changed it to write out 0.0

**Comment by megies** [comment]
see #3225 for last minute changes for release
